File: ios/Packages/PadelCore/Sources/PadelCore/Analytics/PersonalRecords.swift
import Foundation

/// Un récord personal: el valor y la sesión en la que cayó.
public struct PersonalRecord: Equatable, Sendable {
    public let valor: Float
    public let sessionId: String
    public let startedAtEpochMs: Int64
}

/// Los récords del jugador sobre su historial: pura motivación, cero coste de captura.
/// Se recalculan de las sesiones cada vez — como el nivel, no pueden desincronizarse.
/// Espejo del core Kotlin, con tests allí.
public struct PersonalRecords: Equatable, Sendable {
    /// Pala más rápida (km/h).
    public let velocidadMax: PersonalRecord?
    /// Más golpeos en una sesión.
    public let golpeosMax: PersonalRecord?
    /// Mejor ritmo (golpeos/minuto) en sesiones de al menos 10 minutos.
    public let ritmoMax: PersonalRecord?
    /// Mejor nivel de sesión (solo sesiones con nivel fiable).
    public let nivelMax: PersonalRecord?
    /// Más smashes en una sesión.
    public let smashesMax: PersonalRecord?

    /// Sesiones muy cortas fuera del ritmo: 3 golpeos en 1 min no son un récord.
    public static let minRitmoDurationS: Int64 = 10 * 60

    /// ¿Esta sesión ostenta alguno de los récords? Para el distintivo 🏆 en su ficha.
    public func esDe(_ sessionId: String) -> Bool {
        [velocidadMax, golpeosMax, ritmoMax, nivelMax, smashesMax]
            .compactMap { $0 }
            .contains { $0.sessionId == sessionId }
    }

    public static func from(_ sessions: [PadelSession]) -> PersonalRecords {
        func mejor(_ valorDe: (PadelSession) -> Float?) -> PersonalRecord? {
            sessions.compactMap { s -> PersonalRecord? in
                guard let valor = valorDe(s), valor > 0 else { return nil }
                return PersonalRecord(
                    valor: valor, sessionId: s.sessionId, startedAtEpochMs: s.startedAtEpochMs
                )
            }
            .max { $0.valor < $1.valor }
        }

        return PersonalRecords(
            velocidadMax: mejor { $0.intensity.maxRacketSpeedKmh },
            golpeosMax: mejor { Float($0.totalShots) },
            ritmoMax: mejor { s in
                s.durationSeconds >= Self.minRitmoDurationS ? s.shotsPerMinute : nil
            },
            nivelMax: mejor { s in
                let level = s.level
                return level.gradedShots > 0 && level.reliable ? level.overall : nil
            },
            smashesMax: mejor { s in
                s.shotsByType[.smash].map(Float.init)
            }
        )
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/analytics/PersonalRecords.kt
package com.risingpadel.core.analytics

import com.risingpadel.core.model.PadelSession
import com.risingpadel.core.model.ShotType

/**
 * Un récord personal: el valor y la sesión en la que cayó.
 */
data class PersonalRecord(
    val valor: Float,
    val sessionId: String,
    val startedAtEpochMs: Long,
)

/**
 * Los récords del jugador sobre su historial: pura motivación, cero coste de captura.
 * Se recalculan de las sesiones cada vez — como el nivel, no pueden desincronizarse.
 */
data class PersonalRecords(
    /** Pala más rápida (km/h). */
    val velocidadMax: PersonalRecord?,
    /** Más golpeos en una sesión. */
    val golpeosMax: PersonalRecord?,
    /** Mejor ritmo (golpeos/minuto) en sesiones de al menos 10 minutos. */
    val ritmoMax: PersonalRecord?,
    /** Mejor nivel de sesión (solo sesiones con nivel fiable). */
    val nivelMax: PersonalRecord?,
    /** Más smashes en una sesión. */
    val smashesMax: PersonalRecord?,
) {
    /** ¿Esta sesión ostenta alguno de los récords? Para el distintivo 🏆 en su ficha. */
    fun esDe(sessionId: String): Boolean = listOfNotNull(
        velocidadMax, golpeosMax, ritmoMax, nivelMax, smashesMax
    ).any { it.sessionId == sessionId }

    companion object {
        /** Sesiones muy cortas fuera del ritmo: 3 golpeos en 1 min no son un récord. */
        const val MIN_RITMO_DURATION_S = 10 * 60L

        fun from(sessions: List<PadelSession>): PersonalRecords {
            fun mejor(valorDe: (PadelSession) -> Float?): PersonalRecord? =
                sessions.mapNotNull { s ->
                    valorDe(s)?.takeIf { it > 0 }?.let {
                        PersonalRecord(it, s.sessionId, s.startedAtEpochMs)
                    }
                }.maxByOrNull { it.valor }

            return PersonalRecords(
                velocidadMax = mejor { it.intensity.maxRacketSpeedKmh },
                golpeosMax = mejor { it.totalShots.toFloat() },
                ritmoMax = mejor { s ->
                    s.shotsPerMinute.takeIf { s.durationSeconds >= MIN_RITMO_DURATION_S }
                },
                nivelMax = mejor { s ->
                    s.level.takeIf { it.gradedShots > 0 && it.reliable }?.overall
                },
                smashesMax = mejor { s ->
                    s.shotsByType[ShotType.SMASH]?.toFloat()
                },
            )
        }
    }
}
